File: tests/pcsxrunner/web_api.cc
/***************************************************************************
 *   Copyright (C) 2026 PCSX-Redux authors                                 *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.           *
 ***************************************************************************/

// Regression coverage for the -no-ui web/API surface PSoXide-2 depends on.
//
// PSoXide-2 drives Redux in -no-ui mode and pulls trusted-reference frames
// via GET /api/v1/gpu/vram/raw. This test exercises that exact end-to-end
// chain, inside a single pcsx-redux-tests binary:
//
//   MainInvoker (-no-ui -softgpu -interpreter -testmode)
//     -> PCSX.startWebServer(18484)              [Lua binding wraps
//                                                 g_emulator->m_webServer
//                                                 ->startServer(loop, port)]
//     -> luv TCP connect to 127.0.0.1:18484
//     -> GET /api/v1/gpu/vram/raw
//     -> WebServer accept -> WebClient -> llhttp parser -> findExecutor
//        -> VramExecutor::execute -> m_gpu->getVRAM() (SoftGPU m_vram16)
//     -> 1 MiB body + Content-Length: 1048576
//     -> PCSX.quit(0) on success / PCSX.quit(1) on any mismatch
//
// Failure modes this catches:
//   - web server failing to bind / accept under -no-ui (loop ownership,
//     SettingsLoaded bypass, TCP setup regressions)
//   - VramExecutor path regressing (route match, 200 framing, body size)
//   - soft-GPU getVRAM() shape regressing (not 1024*512*2 bytes)
//   - llhttp / uriparser / executor dispatch regressions on the request path
//   - the Lua startWebServer binding regressing (direct PSoXide-2 dep)
//
// Port 18484 is non-default (Redux default is 8080) to avoid colliding with
// any dev-local Redux instance a contributor may have running. No -luacov:
// same rationale as the other no-ui regressions in this directory — the
// path under test is C++/Lua only, and luacov would write into the repo
// root at shutdown, which is noise in sandboxed CI.

#include "gtest/gtest.h"
#include "main/main.h"

TEST(WebAPI, NoUIVramRawGet) {
    MainInvoker invoker("-no-ui", "-bios", "src/mips/openbios/openbios.bin", "-testmode", "-interpreter", "-softgpu",
                        "-dofile", "tests/pcsxrunner/web_api.lua");
    int ret = invoker.invoke();
    EXPECT_EQ(ret, 0);
}

File: tests/pcsxrunner/web_api.lua
-- Regression driver for WebAPI.NoUIVramRawGet.
--
-- Brings up the web server under -no-ui and exercises the exact capture
-- endpoint PSoXide-2's trusted-reference pipeline calls to pull frames.
-- The request goes through the real TCP path, the HTTP parser, the
-- VramExecutor, and g_emulator->m_gpu->getVRAM() on the soft GPU.

local port = 18484

PCSX.startWebServer(port)

local client = luv.new_tcp()
local timer = luv.new_timer()
local buf = {}
local done = false

local function finish(code, msg)
    if done then return end
    done = true
    pcall(function() luv.timer_stop(timer); luv.close(timer) end)
    pcall(function() luv.close(client) end)
    if msg then io.stderr:write('WebAPI failure: ' .. msg .. '\n') end
    PCSX.quit(code)
end

luv.timer_start(timer, 5000, 0, function() finish(1, 'timed out waiting for response') end)

luv.tcp_connect(client, '127.0.0.1', port, function(connectErr)
    if connectErr then finish(1, 'connect: ' .. tostring(connectErr)); return end
    luv.write(client, 'GET /api/v1/gpu/vram/raw HTTP/1.1\r\nHost: localhost\r\n\r\n')
    luv.read_start(client, function(readErr, chunk)
        if readErr and readErr ~= 'EOF' then finish(1, 'read: ' .. tostring(readErr)); return end
        if chunk then
            buf[#buf + 1] = chunk
            return
        end
        -- EOF: validate
        local full = table.concat(buf)
        local headerEnd = full:find('\r\n\r\n', 1, true)
        if not headerEnd then finish(1, 'no header terminator; got ' .. #full .. ' bytes'); return end
        local headers = full:sub(1, headerEnd - 1)
        local body = full:sub(headerEnd + 4)
        local status = headers:match('^HTTP/1%.1 (%d+)')
        local clen = tonumber(headers:match('[Cc]ontent%-[Ll]ength:%s*(%d+)'))
        if status ~= '200' then finish(1, 'status ' .. tostring(status)); return end
        if clen ~= 1048576 then finish(1, 'content-length ' .. tostring(clen)); return end
        if #body ~= 1048576 then finish(1, 'body size ' .. #body .. ' != 1048576'); return end
        finish(0, nil)
    end)
end)
